### ISSUE TITLE
[ban-types] allow `extends Function` by default?

### ISSUE THREAD

**Issue by thw0rted** [typescript-eslint/typescript-eslint#4390]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    // allow default config for "@typescript-eslint/ban-types"
  }
}
```

```TS
type WithoutMethods<T> = {
  [K in keyof T]: T[K] extends Function ? never : T[K]
}
```

**Expected Result**

Don't flag the use of `Function` in the context of a conditional or mapped type, where it is useful to know that a type is callable, but its specific call signature is definitely irrelevant.

**Actual Result**

The subject rule flags this use of `Function`.

**Additional Info**

This was raised in https://github.com/typescript-eslint/typescript-eslint/pull/848#issuecomment-569327935 but nobody in that thread addressed it, and the thread is now locked.  (Somebody also brought up "extends object"; one could argue that this should also be allowed for similar reasons, but IMHO "extends Function" is slightly more important.)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.7.0` |
| `@typescript-eslint/parser`        | `5.7.0` |
| `TypeScript`                       | `4.5.4` |
| `ESLint`                           | `8.4.1` |

**Comment by bradzacher** [comment]
`ban-types` simply looks for "banned" names. It does not care about context because it cannot - it is fully configurable. We don't want the rule to behave differently for specific names that you might pass in as that would violate the law of least surprise. It would be confusing and surprising that `Function` can be used in an `extends` even if it's banned by `ban-types`, but any other type would not behave in the same way.

Disable comments exist for this purpose and are much clearer than adding special handling to the rule because you can document why you're ignoring the rule so nobody needs to learn about the surprising behaviour.
```ts
type WithoutMethods<T> = {
  // eslint-disable-next-line @typescript-eslint/ban-types -- this is safe as we only care that the type is callable and we never call it.
  [K in keyof T]: T[K] extends Function ? never : T[K]
}
```

**Comment by thw0rted** [comment]
I appreciate that it isn't compatible with the way the rule is implemented, so it doesn't make sense to change, but I disagree that it would violate POLS.  I look at linter rules as a way of codifying best practices; in this case, typing variables or parameters as `Function` should be avoided, but using `extends Function` in a conditional type should be encouraged, as it's clearer than substituting a specific call signature.

The best rule documentation pages are the ones that explain what is and isn't allowed, and *why*.  The community could discuss and agree on the "right" way to test callability, then document it.  Say it's a limitation of the rule engine, but `extends Function` is sound, and should be manually marked as you describe.  As it stands, I'm "surprised" when the linter flags something I thought I was doing right, and it turns into a research project to see if there's a different syntax I'm "supposed" to use (and what corner-case will bite me if I don't).

**Comment by bradzacher** [comment]
The reason you're surprised is due to how you're thinking about the rule.
From what you're saying - it looks like you're thinking of this rule as "don't use these types except when it's safe to use them".
But it's not that - instead it's just "don't use these types, period".

With the former definition - it is surprising that it doesn't allow "safe" usages.
With the latter definition - it is ***un***surprising, and allowing "safe" usages would be surprising.

----

As another example. By default `{}` is banned because most people think it represents "an object type with no properties". However, this is incorrect - it represents "any non-nullish type".

In frameworks like react this is problematic because it's very common to declare a component with prop type `{}` - meaning "this component accepts no properties".
On the surface this looks unsafe - your component could accept a string instead of an object as its props! However due to how the react types are written, this prop type is intersected with an actual object with properties before being exposed to JSX - meaning the unsafety is "removed" (because `{a: string} & {} === {a: string}`).

But how can you encode that in the lint rule? How can you teach it that this particular case is safe? This is way more complex than your usecase for sure, but the essence of it is the same - "why does ban types ban this type when it's safe to use here?"
The answer is the same - because it just bans the type, period.

----

There's probably space in the ecosystem for a rule individually built for each of the types like `Function`, `Object`, `{}` and `object` which bans them unless it's "safe" to use them - but this rule is not it.
